scripts/plugins/extauth-hook-AD.py

```diff
@@ -9,138 +9,304 @@
-        name_lines = filter(lambda x: x.startswith("Name:"), lines)
-        if len(name_lines) != 1:
-            # Just warn, don't raise exception - there may be others that work
-            log_err("Could not find user/group corresponding to SID %s (%s)" % (sid,str(lines)))
-            return
-        name = name_lines[0].split(":")[1].strip()
-        uid_lines = filter(lambda x: x.startswith("Uid:"), lines)
-        if len(uid_lines) == 0:
-            is_group = True
-        else:
-            is_group = False
+session    sufficient     pam_lsass.so"""
+
+    def __init__(self, session, args, ad_enabled=True):
+        super(StaticPam, self).__init__("/etc/pam.d/sshd", session, args, ad_enabled,
+                                        load_existing=False)
 
-        if is_group:
-            os.write(self.temp_fd, "account sufficient pam_succeed_if.so user ingroup %s\n" % name)
+    def _apply_to_cache(self):
+        if self._ad_enabled:
+            if self._backend == ADBackend.bd_pbis:
+                ad_pam_module = "/lib/security/pam_lsass.so"
+            else:
+                ad_pam_module = "pam_winbind.so"
+            content = self.ad_pam_format.format(ad_module=ad_pam_module)
         else:
-            os.write(self.temp_fd, "account sufficient pam_succeed_if.so user = %s\n" % name)
+            content = self.no_ad_pam
+        self._lines = content.split("\n")
 
+class DynamicPam(ADConfig):
+    def __init__(self, session, arg, ad_enabled=True):
+        super(DynamicPam, self).__init__("/etc/pam.d/hcp_users", session, arg, ad_enabled,
+                                         load_existing=False)
 
-def rewrite_etc_pamd_ssh(session, args):
-    # Rewrite the PAM SSH config using the latest info from Active Directory
-    # and the list of subjects from xapi
-    try:
-        config = PamSshConfig()
-        subjects = session.xenapi.subject.get_all()
-        admin_role = session.xenapi.role.get_by_name_label('pool-admin')[0]
+    def _apply_to_cache(self):
+        # AD is not enabled, just return as the configure file will be removed during install
+        if not self._ad_enabled:
+            return
+        # Rewrite the PAM SSH config using the latest info from Active Directory
+        # and the list of subjects from xapi
+        subjects = self._session.xenapi.subject.get_all()
+        admin_role = self._session.xenapi.role.get_by_name_label('pool-admin')[0]
         # Add each subject which contains the admin role
         for opaque_ref in subjects:
-            subject_rec = session.xenapi.subject.get_record(opaque_ref)
+            subject_rec = self._session.xenapi.subject.get_record(opaque_ref)
             sid = subject_rec['subject_identifier']
+            name = subject_rec["other_config"]["subject-name"]
+            is_group = subject_rec["other_config"]["subject-is-group"] == "true"
+            logger.debug("Permit %s with sid %s is_group as %s", name, sid, is_group)
```

Comment:
Shouldn't this logline be part of the if statement?

Reply:
oh, it should be, nice catch